### PR TITLE
Unify DataLoader APIs

### PR DESCRIPTION
**Pull request by sneaxiy**
See Chinese doc [here](https://github.com/PaddlePaddle/FluidDoc/pull/1162).

This PR unifies `PyReader` and `Dataset` interfaces to `DataLoader.from_xxx` API. Users can create different `DataLoader` object to load data, which can be used by:
```
# from generator
data_loader = fluid.io.DataLoader.from_generator(feed_list=[image, label], capacity=16)
                                .set_batch_generator(user_defined_reader, places=fluid.cuda_places())

# from dataset
data_loader = fluid.io.DataLoader.from_dataset(dataset, places=fluid.cpu_places())

# for-range iterations
for epoch_id in range(EPOCH_NUM):
    for data in data_loader():
         exe.run(feed=data, ...)
```

Document Preview:
`DataLoader.from_generator`:
image

`DataLoader.from_dataset`:
image

`PyReader.next`:
image

### CODE REVIEW

File: python/paddle/fluid/reader.py [hunk 1]
```diff
@@ -44,7 +47,354 @@ def _convert_places(places):
...
+        self._places = None
+        self._thread = None
+        self._feed_list = feed_list
+        if not capacity:
+            raise ValueError("Please give value to capacity.")
+        # force to use iterable mode under dygraph mode
+        if in_dygraph_mode():
+            if not iterable:
+                warnings.warn(
+                    "Please NOTE: dygraph can support iterable mode only.")
```
Comment (zhhsplendid) [inline python/paddle/fluid/reader.py]: Had better add "Change to iterable mode".

File: python/paddle/fluid/reader.py [hunk 2]
```diff
@@ -44,7 +47,354 @@ def _convert_places(places):
...
+            self._iterable = True
+            if not return_list:
+                warnings.warn(
+                    "Please NOTE: dygraph can support return as list only.")
```
Comment (zhhsplendid) [inline python/paddle/fluid/reader.py]: Had better add "Change to return as list".

Comment (sneaxiy) [reply python/paddle/fluid/reader.py]: Done.

File: python/paddle/fluid/reader.py (hunk 1 above)
Comment (sneaxiy) [reply python/paddle/fluid/reader.py]: Done.

**Review by chengduoZH — DISMISSED** [review]

**Review by lanxianghit — DISMISSED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by wanghaoshuang — APPROVED** [review]
LGTM on API

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**